python/cudf/cudf/core/groupby/groupby.py

```diff
@@ -582,11 +700,176 @@ def rank(x):
+        if not plc_columns:
+            plc_table = None
+        else:
+            plc_table = plc.Table(plc_columns)
+        offsets, grouped_keys, grouped_values = (
+            self._groupby.plc_groupby.get_groups(plc_table)
         )
 
+        return (
+            offsets,
+            [ColumnBase.from_pylibcudf(col) for col in grouped_keys.columns()],
+            (
+                [
+                    ColumnBase.from_pylibcudf(col)
+                    for col in grouped_values.columns()
+                ]
+                if grouped_values is not None
+                else []
+            ),
+        )
+
+    def _aggregate(
+        self, values: tuple[ColumnBase, ...], aggregations
+    ) -> tuple[
+        list[list[ColumnBase]],
+        list[ColumnBase],
+        list[list[tuple[str, str]]],
+    ]:
+        included_aggregations = []
+        column_included = []
+        requests = []
+        result_columns: list[list[ColumnBase]] = []
+        for i, (col, aggs) in enumerate(zip(values, aggregations)):
+            valid_aggregations = get_valid_aggregation(col.dtype)
+            included_aggregations_i = []
+            col_aggregations = []
+            for agg in aggs:
+                str_agg = str(agg)
+                if (
+                    is_string_dtype(col)
+                    and agg not in _STRING_AGGS
+                    and (
+                        str_agg in {"cumsum", "cummin", "cummax"}
+                        or not (
+                            any(
+                                a in str_agg
+                                for a in {
+                                    "count",
+                                    "max",
+                                    "min",
+                                    "first",
+                                    "last",
+                                    "nunique",
+                                    "unique",
+                                    "nth",
+                                }
+                            )
+                            or (agg is list)
+                        )
+                    )
+                ):
+                    raise TypeError(
+                        f"function is not supported for this dtype: {agg}"
+                    )
+                elif (
+                    _is_categorical_dtype(col)
+                    and agg not in _CATEGORICAL_AGGS
+                    and (
+                        str_agg in {"cumsum", "cummin", "cummax"}
+                        or not (
+                            any(
+                                a in str_agg
+                                for a in {"count", "max", "min", "unique"}
+                            )
+                        )
+                    )
+                ):
+                    raise TypeError(
+                        f"{col.dtype} type does not support {agg} operations"
+                    )
```

Comment:
Would you be okay with moving this logic out into separate functions that determine if an aggregation is not supported for string and categorical types? They can be nested in `_aggregate`. Eg.
```python
def _is_unsupported_for_string(col, str_agg):
    cumulative_agg = str_agg in {"cumsum", "cummin", "cummax"}
    basic_agg = any(a in str_agg for a in {
        "count", "max", "min", "first", "last", "nunique", "unique", "nth"
    })
    return (
        is_string_dtype(col)
        and str_agg not in _STRING_AGGS
        and (cumulative_agg or (not basic_agg and str_agg != "list"))
    )

def _is_unsupported_for_categorical(col, str_agg):
    cumulative_agg = atr_agg in {"cumsum", "cummin", "cummax"}
    basic_agg = any(a in str_agg for a in {"count", "max", "min", "unique"})
    return (
        _is_categorical_dtype(col)
        and agg_str not in _CATEGORICAL_AGGS
        and (cumulative_agg or not basic_agg)
    )
```

Reply:
Maybe use a `singledispatch` function on the dtype.

Reply:
Sure thing. Was able adapt this to use `singledispatch`